torch/testing/_internal/common_methods_invocations.py

```diff
@@ -6098,6 +6113,11 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
            supports_inplace_autograd=True,
            safe_casts_outputs=True,
            sample_inputs_func=sample_inputs_xlogy),
+    OpInfo('zero_',
+           dtypes=all_types_and_complex_and(torch.bfloat16, torch.half, torch.bool),
+           supports_autograd=False,
```

Comment:
This should support autograd.
https://github.com/pytorch/pytorch/blob/6de11c2db37d2b64debb97af40147de1581f718a/tools/autograd/derivatives.yaml#L1274

Reply:
This is the error I got for almost all the tests:

```
RuntimeError: a leaf Variable that requires grad is being used in an in-place operation.
```

Do you have any idea what is the solution to this? @kshitij12345

Reply:
Do you still get an error after wrapping the operator in lambda?

Reply:
I think the tests always assume that the operator's function variant (which was always true except for this operator) doesn't modify the input inplace.

eager consistency test has some mechanism to deal with this,

https://github.com/pytorch/pytorch/blob/d88d321ee3a65894baf6dc0a580b44180ec52f3c/test/test_ops.py#L321

Need to dig a bit further.

Reply:
REG: https://github.com/pytorch/pytorch/pull/58731#discussion_r636674476

Yes, still getting the same errors after wrapping the operator in lambda.

Reply:
Inspired by 
https://github.com/pytorch/pytorch/blob/d88d321ee3a65894baf6dc0a580b44180ec52f3c/test/test_ops.py#L321

I tried,
```python
OpInfo('zero',
       op=lambda x: torch.zero_(x.clone()),
       dtypes=all_types_and_complex_and(torch.bfloat16, torch.half, torch.bool),
       supports_autograd=True,
       supports_out=False,
       sample_inputs_func=sample_inputs_zero_),
```

which works for all test except `jit` and `eager` consistency.
`jit` consistency has issues with op being wrapped in lambda.

The eager test fails for more subtle reasons.

https://github.com/pytorch/pytorch/blob/d88d321ee3a65894baf6dc0a580b44180ec52f3c/torch/testing/_internal/common_methods_invocations.py#L224-L228

We get the `method` variant and `inplace` variant in the above way.

For this case, `method` variant is `Tensor.zero_` and inplace variant is `None`.

Thus in eager tests at the point below,
https://github.com/pytorch/pytorch/blob/d88d321ee3a65894baf6dc0a580b44180ec52f3c/test/test_ops.py#L321-L323

We incorrectly don't clone the inputs for method variant (which is actually the inplace variant) and hence we get,

```python
RuntimeError: a leaf Variable that requires grad is being used in an in-place operation.
```

To mitigate this, I propose to allow setting `method_variant` and `inplace_variant` via the OpInfo constructor.

cc: @mruberry

Reply:
Thanks, @kshitij12345 for digging into this. Taking a look at `test/test_autograd.py` to understand how autograd is tested for `method_tests`, it seems like `requires_grad` is set to `False` (/reset) for inplace ops. There is an additional comment that mentions that it needs to be fixed. Please see the relevant snippet below.

```python
is_magic_method = name[:2] == '__' and name[-2] == '__'
is_inpalce = name[-1] == "_" and not is_magic_method
self_variable = create_input((self_size,), dtype=dtype, device=device)[0][0]
# FixMe: run grad checks on inplace self
if is_inplace:
    self_variable.requires_grad = False
```

Reference: https://github.com/pytorch/pytorch/blob/master/test/test_autograd.py#L5474

Reply:
For the scope of this PR, I suggest to jit and eager consistency checks, and we can file this an issue (on allowing to set `method_variant` and `inplace_variant` via OpInfo constructor) separately and take the discussion there.

Looking forward to know your views, @kshitij12345, @mruberry.

Reply:
Skipping those tests and following up by allowing method and inplace variants to be specified sounds good